NB-4 feature

JCameronBath opened the pull request:
Added a new switch case for adding a new account
Added a new method for the creation of that new account
Imported scanner class

JCameronBath committed:
Added a new switch case for adding a new account
Added a new method for the creation of that new account
Imported scanner class

JCameronBath committed:
Added a new switch case for adding a new account
Added a new method for the creation of that new account
Imported scanner class

marklawrence97 commented on newbank/server/NewBank.java:
```diff
@@ -46,6 +48,7 @@ public synchronized String processRequest(CustomerID customer, String request) {
 		if(customers.containsKey(customer.getKey())) {
 			switch(request) {
 			case "SHOWMYACCOUNTS" : return showMyAccounts(customer);
+			case "ADDNEWACCOUNT" :  return addNewAccount(customer);
```
I think in the protocol (see, (https://github.com/marklawrence97/NewBank/blob/master/protocol.txt)) the string to enter to create a new account is simply NEWACCOUNT - not sure how closely we need to stick to this protocol, potentially it would be helpful to list the currently available options to the user when we ask what they want and what they need to enter, then we can diverge?

marklawrence97 commented on newbank/server/NewBank.java:
```diff
@@ -1,9 +1,11 @@
 public class NewBank {
-	
+
+	Scanner reader = new Scanner(System.in);
```
This scanner will listen for input and output on the server instead of the client and so unfortunately it is not possible to interact with the user in this manner. 

NewBankClientHandler has properties (out and in) that allow for this interaction. The only thing that gets printed to the client from this currently is "Account has been created", which is returned by addNewAccount and then printed to the client. 

I think for now the first iteration of this feature it would be acceptable to simple follow the protocol and allow the user to enter NEWACCOUNT <name>, where <name> is the name of the account and then return either SUCCESS or FAIL

JCameronBath replied:
I'm planning to allow it just for the 3 types of accounts listed in the code (Main, savings, and checking), potentially with us following up in the new week on whether other types can be allowed. As for listing the options, perhaps this could also be a follow up to the customer next week? Seems like it may get kind of clunky as we add in the new options, but would be helpful to clarify before adding it.

Does that seem okay to you? Essentially, update the name and set it up for the accounts we know can exist.

marklawrence97 replied:
Yes I agree I think listing options would be better to be done on a later ticket after discussion, I think for now then do you think it would be worth updating the protocol.txt file with the new command so a user doesn't need to go through the code find out how to create an account?

I do think it would make more sense to have all the interaction with the user on the client instead of the server as this saves the user having to flick between the client and the server !

JCameronBath replied:
Have updated the code to allow for these accounts (failing if already exists, since more than 1 of the same would make things like transfer have issues). Also updated the protocol.txt file to mention the possible names currently.

marklawrence97 replied:
Apologies - I think I may have been going through an old commit so my comments may have been outdated, will go through the newer commits. 

Not sure how everyone else feels about git history so this is more of an opinion / one for the team but instead of having a commit for each amendment would it make more sense to have one commit per feature / change. 

So instead of creating a new commit with the same name the new changes can be amended to the latest commit with the command:

`git commit --amend --no-edit`

Amending the commit has to be done when you have staged changes you want to commit. This can also be achieved retrospectively by squashing the commits, this can be done with an interactive rebase. For example you could use:

`git rebase -i develop`

I would suggest a few of the commits could be squashed into one by selecting picking the most recent and selecting 'fixup' for the others with the same name.

Definitely not required for approval but potentially useful to bear in mind !

JCameronBath committed:
Added new switch cases for adding a new account
Added corresponding new methods for the creation of that new account

JCameronBath committed:
Added new switch cases for adding a new account
Added corresponding new methods for the creation of that new account
Returns fail if already exists for the customer

JCameronBath committed:
Update protocol.txt

Added the possible account names currently

marklawrence97 commented on newbank/server/NewBank.java:
```diff
@@ -46,14 +46,45 @@ public synchronized String processRequest(CustomerID customer, String request) {
+			return "FAIL";
+		} else {
+			customers.get(customer.getKey()).addAccount(new Account("Main", 0.0));
+			return "SUCCESS";
```
Do we definitely have to restrict to savings/main/checking and not just allow a user to specify an arbitrary name for the account?

Potentially savings/main/checking could be an account type in a later ticket?

Happy for you to decide on this, these functions could be condensed into one with a second argument that contains the account name, for example:

```
addAccount(CustomerID customer, string Name) {
   // same logic but replace the string "Main" with the variable name
}
```

Then happy to approve !

JCameronBath replied:
I've asked Nadi to check with the customer on whether to allow arbitrary or only these specific names (as well as the question from earlier about listing the possible requests so they don't need to flick back and forth between client and protocol).
Thought I'd put it in like this for now as a sort of filler, and then can generalise if wanted by the customer

marklawrence97 replied:
Awesome makes sense, sounds great !

JCameronBath committed:
Added new switch cases for adding a new account
Added corresponding new method for the creation of that new account
Returns fail if already exists for the customer

JCameronBath committed:
Merge remote-tracking branch 'origin/feature' into feature

JCameronBath committed:
Update protocol.txt

JCameronBath commented:
Think we should just settle with this for now, and then a new pull request can be done later if needed. Would be good to get basic functionalities in for now

marklawrence97 commented:
I agree, I will approve this for now but I think we should add a ticket for allowing user to input a custom name for the account.

This PR provides the basic functionality outlined by the ticket so I think it is fine to merge this.

marklawrence97 reviewed (approved):
Currently all accounts will have to be called `"<name>"` but this can be addressed in a follow up PR, this provides all the functionality required
